proposals/extensions_v2.md

```diff
@@ -0,0 +1,371 @@
+
+These areas are:
+
+## Expansion 1: Syntactic clumsiness and repetition
+
+The initial extension form considers source and binary compatibility as core requirements that must be present to ensure easy migration, allowing codebases to avoid both:
+1. bifurcation; where some codebases adopt modern extensions and some do not.
+2. internal inconsistency; where some codebases must keep around old extensions and new extensions, with confusion about the semantics of how each interacts with the other.
+
+Because classic extension methods have very few restrictions, modern extension methods need to be flexible enough to support all the scenarios which they support.
+
+However, many codebases do not need all the flexibility that classic extension methods afforded.  For example, classic extension methods allow disparate extension methods in a single static class to target multiple different types.  For use cases where that isn't required, we forsee a natural extension (pun intended) where one can translate a modern extension like so:
+
+```c#
+extension E
+{
+    // All extension members extend the same thing:
+
+    public void M() for SomeType { ... }
+    public int P for SomeType { get { ... } }
+    public static operator+(...) for SomeType { ... }
+    // etc
+}
+
+// Can be translated to:
+
+extension E for SomeType
+{
+    public void M() { ... }
+    public int P { get { ... } }
+    public static operator+(...) { ... }
+}
+```
+
+TODO: Do an ecosystem check on what percentage of existing extensions could use this simpler form.
+
+TODO: It's possible someone might have an extension where almost all extensions extend a single type, and a small handful do something slightly different (perhaps extending by `this ref`).  Would it be beneficial here to *still* allow the extension members to provide a `for-clause` to override that default for that specific member.  For example:
+
+```c#
+extension StringExtensions for string str
+{
+    // Lots of normal extension methods on string ...
+
+    // Override here to extend `string?`
+    public bool MyIsNullOrEmpty() for [NotNullWhen(false)] string? str
+    {
+    }
+}
+```
+
+It seems like this would be nice to support with little drawback.
+
+## Expansion 2: Optional syntactic components
+
+As above, we want modern extensions to completely subsume classic extension methods.  As such, a modern extension  method must be able to support everything a classic extension method supported.  For example:
+
+```c#
+static class Extensions
+{
+    // Yes, this is legal
+    public static void MakeNonNull([Attr] this ref int? value)
+    {
+        if (value is null)
+            value = 0;
+    }
+}
+```
+
+For this reason, the strawman syntax for this `for-clause` is `'for' parameter`, where `parameter` is the familiar:
+
+```g4
+parameter
+    | attributes? modifiers? type identifier
+    | attributes? modifiers? type identifier '=' expression
+    ;
+```
```

Comment:
````suggestion
For this reason, the strawman syntax is:

```g4
for-clause
    | 'for' parameter`
    ;

parameter (unchanged)
    | attributes? modifiers? type identifier
    | attributes? modifiers? type identifier '=' expression
    ;
```
````